### PR TITLE
Added one-line function initializeKeypair for quick startup for dev labs

### PR DESCRIPTION
**Pull request by CoachChuckFF**
Adding new function: initializeKeypair
 * Loads a keypair from the environment or generates a new one if not found and writes it to the environment,
 * then requests an airdrop if the balance is below the minimum balance

### CODE REVIEW

**Review by nickfrosty — APPROVED** [review]
Nice addition Christian (and yes, I rhymed on purpose) :)

**Comment by aitimate** [comment]
```ts
import {Keypair} from "@solana/web3.js";
import * as fs from "fs"
import "dotenv/config";

export function initializeKeypair(name: string, keypair?: Keypair) {
    if (!process.env[name]) {
        keypair ??= Keypair.generate()
        fs.appendFileSync(".env", `\n# ${keypair.publicKey}\n${name}=[${keypair.secretKey}]`);
    }
    return keypair ?? Keypair.fromSecretKey(Uint8Array.from(JSON.parse(process.env[name]!)))
}
```

**Comment by mikemaccana** [comment]
@aitimate thanks for the comment, but that would duplicate code that's already elsewhere in the library and not fund the resulting keypairs.

File: src/index.test.ts [hunk 1]
```diff
@@ -168,7 +167,7 @@ describe("addKeypairToEnvFile", () => {
 
     assert.ok(envKeypair.secretKey);
 
-    deleteFile(".env");
+    await deleteFile(".env");
```
Comment (mikemaccana) [inline src/index.test.ts:170]: Good spot.

File: src/index.test.ts [hunk 2]
```diff
@@ -182,6 +181,49 @@ describe("addKeypairToEnvFile", () => {
   });
 });
 
+describe("initializeKeypair", () => {
```
Comment (mikemaccana) [inline src/index.test.ts:184]: Sweet let's get this is now. It looks great BTW, reloading and checking keypair etc is exactly how I'd do the tests.

Comment (mikemaccana) [reply src/index.test.ts:184]: Can you update the README.md with a description, example etc. per the other items?

File: src/index.ts [hunk 3]
```diff
@@ -149,6 +155,61 @@ export const addKeypairToEnvFile = async (
   await appendFile(fileName, `\n${variableName}=${secretKeyString}`);
 };
 
+export interface initializeKeypairOptions {
+  envFileName?: string; // .env file path
```
Comment (mikemaccana) [inline src/index.ts:159]: Please ditch the Javadoc-style comments. I think I mentioned this elsewhere but the idea is we use types to determine types, whether something is optional, etc. Use comments for situations that aren't obvious to a TS developer. This increases the signal:noise ratio of the code and reduces the chance of out of date comments misleading other developers.

File: src/index.ts [hunk 4]
```diff
@@ -149,6 +155,61 @@ export const addKeypairToEnvFile = async (
...
+    signer = getKeypairFromEnvironment(variableName);
+  } else {
+    signer = Keypair.generate();
+    await addKeypairToEnvFile(signer, variableName, envFileName);
+  }
+
+  await airdropIfRequired(
+    connection,
+    signer.publicKey,
+    airdropAmount || LAMPORTS_PER_SOL,
```
Comment (mikemaccana) [inline src/index.ts:206]: Can you make this explicit?

```
const DEFAULT_AIRDROP_AMOUNT = 1 * LAMPORTS_PER_SOL;

...

airdropAmount || DEFAULT_AIRDROP_AMOUNT
```

File: src/index.ts [hunk 5]
```diff
@@ -149,6 +155,61 @@ export const addKeypairToEnvFile = async (
...
+    minimumBalance || LAMPORTS_PER_SOL >> 1,
```
Comment (mikemaccana) [inline src/index.ts:207]: Using [Bitwise shift operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift) wouldn't be apparent to most TS devs (knowing the state of TS devs lol) and I'm not sure there's a particular reason to use the bitwise shift operator.

```
const DEFAULT_MINIMUM_BALANCE = 500_000_000
```

File: src/index.ts [hunk 6]
```diff
@@ -157,7 +218,7 @@ const requestAndConfirmAirdrop = async (
   publicKey: PublicKey,
   amount: number,
 ) => {
-  let airdropTransactionSignature = await connection.requestAirdrop(
+  const airdropTransactionSignature = await connection.requestAirdrop(
```
Comment (mikemaccana) [inline src/index.ts:221]: Good spot!

File: src/index.ts [hunk 7]
```diff
@@ -149,6 +155,61 @@ export const addKeypairToEnvFile = async (
...
+): Promise<Keypair> => {
+  let {
+    envFileName,
+    variableName,
+    airdropAmount,
+    minimumBalance,
+    keypairPath,
+  } = options || {};
+
+  let signer: Keypair;
```
Comment (mikemaccana) [inline src/index.ts:191]: Just use the name `keypair`. Our function is called `initializeKeypair` the thing it returns should be `keypair`.

File: src/index.ts (hunk 6 above)
Comment (mikemaccana) [reply src/index.ts:221]: Everything else looks good, get the above sorted and we can get this in today! 🙏

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Thanks @CoachChuckFF we've been [meaning to do this for a while](https://github.com/solana-developers/helpers/issues/12)! A few minor bits but once sorted we can get this in immediately.

File: src/index.ts (hunk 3 above)
Comment (CoachChuckFF) [reply src/index.ts:159]: 10-4

File: src/index.ts (hunk 7 above)
Comment (CoachChuckFF) [reply src/index.ts:191]: Deal

File: src/index.ts (hunk 4 above)
Comment (CoachChuckFF) [reply src/index.ts:206]: Done, DEFAULT_AIRDROP_AMOUNT has been added to the top of the file

File: src/index.ts (hunk 5 above)
Comment (CoachChuckFF) [reply src/index.ts:207]: lol sorry, I come from embedded systems land. Shifts are more efficient than division ( if you want to divide by powers of 2 ). Anyways, I totally agree with you, let's go for readability over me thinking I'm being clever.

File: src/index.test.ts (hunk 2 above)
Comment (CoachChuckFF) [reply src/index.test.ts:184]: Yep!

**Comment by aitimate** [comment]
> @aitimate thanks for the comment, but that would duplicate code that's already elsewhere in the library and not fund the resulting keypairs.



> @aitimate thanks for the comment, but that would duplicate code that's already elsewhere in the library and not fund the resulting keypairs.

In fact, in my development work, I only need to use airdrops occasionally. More often, what I need is to initialize key pairs. This is a problem I encounter in my actual development projects. We don't need to pass an extra connection parameter for the 10% or even less frequent cases. I'm quite sure I understand what you mean.
I sincerely hope you could reconsider the necessity of including the airdrop

**Review by mikemaccana — APPROVED** [review]
Thanks @CoachChuckFF !

**Comment by mikemaccana** [comment]
@aitimate if you just need to make keypairs: https://github.com/solana-developers/helpers?tab=readme-ov-file#make-multiple-keypairs-at-once

**Pull request merged**